### DISCUSSION TITLE
MauiIcons Core Markup Extension Improvements with Setters and Triggers Support

### DISCUSSION THREAD

**Discussion by AathifMahir** [AathifMahir/MauiIcons#106] [General]
This discussion brings together all the feedback on supporting Style Setters and Triggers to Existing MauiIcons Xaml MarkupExtension or Evolution of Xaml Markup Extension Does Fixes the Existing Pitfalls and etc..

**Comment by albyrock87** [comment]
A Markup Extension is not suited for applying values to multiple properties.
It's only purpose is to provide a single value, potentially depending on the injection context.

The current Markup Extension [is trying to do something](https://github.com/AathifMahir/MauiIcons/blob/09016b21e5db46c433405c89cd42e5f6c2c6bdea/src/MauiIcons.Core/Extensions/BaseIconExtension.cs#L93) which exceeds the purpose of a markup extension and in fact this doesn't work on style `Setter`s.

There's also a [legit behavior here](https://github.com/AathifMahir/MauiIcons/blob/09016b21e5db46c433405c89cd42e5f6c2c6bdea/src/MauiIcons.Core/Extensions/BaseIconExtension.cs#L172).

I think we should
- make the markup extension not a markup extension but simply a `BindableObject`, maybe renaming it to `BaseIcon`
- rename the `Icon` [property here](https://github.com/AathifMahir/MauiIcons/blob/09016b21e5db46c433405c89cd42e5f6c2c6bdea/src/MauiIcons.Core/Controls/MauiIcon.cs#L8C5-L8C136) to `IconValue` and make it a private bindable property
- add a new `Icon` property as an attached one of type `MauiIcon` and make use of `propertyChanged:` callback to:
  - bind `BindingContextProperty` in `BaseIcon` to target bindable object `BindingContext` in order to propagate the context
  - then bind all the `BaseIcon` properties to the target ones depending on target type (like we're doing on the extension right now)

This is what it'd look like:
```xml
<mi:MauiIcon Icon="{mi:Cupertino Airplane}"/>
<Label mi:MauiIcon.Icon="{mi:Cupertino Airplane}"/>
```

**Reply by AathifMahir** [reply]
> A Markup Extension is not suited for applying values to multiple properties. It's only purpose is to provide a single value, potentially depending on the injection context.
> 
> The current Markup Extension [is trying to do something](https://github.com/AathifMahir/MauiIcons/blob/09016b21e5db46c433405c89cd42e5f6c2c6bdea/src/MauiIcons.Core/Extensions/BaseIconExtension.cs#L93) which exceeds the purpose of a markup extension and in fact this doesn't work on style `Setter`s.
> 
> There's also a [legit behavior here](https://github.com/AathifMahir/MauiIcons/blob/09016b21e5db46c433405c89cd42e5f6c2c6bdea/src/MauiIcons.Core/Extensions/BaseIconExtension.cs#L172).
> 
> I think we should
> 
> * make the markup extension not a markup extension but simply a `BindableObject`, maybe renaming it to `BaseIcon`
> * rename the `Icon` [property here](https://github.com/AathifMahir/MauiIcons/blob/09016b21e5db46c433405c89cd42e5f6c2c6bdea/src/MauiIcons.Core/Controls/MauiIcon.cs#L8C5-L8C136) to `IconValue` and make it a private bindable property
> * add a new `Icon` property as an attached one of type `MauiIcon` and make use of `propertyChanged:` callback to:
>   
>   * bind `BindingContextProperty` in `BaseIcon` to target bindable object `BindingContext` in order to propagate the context
>   * then bind all the `BaseIcon` properties to the target ones depending on target type (like we're doing on the extension right now)
> 
> This is what it'd look like:
> 
> ```
> <mi:MauiIcon Icon="{mi:Cupertino Airplane}"/>
> <Label mi:MauiIcon.Icon="{mi:Cupertino Airplane}"/>
> ```

This approach looks super interesting, I'll create a POC based on This on a Seperate Branch, I would appriciate if you can provide a Feedback after Looking into that

**Reply by AathifMahir** [reply]
By the way, in the above example, we are still using Markup Extension to Set the Icon?

I mean this one
```xml
<mi:MauiIcon Icon="{mi:Cupertino Airplane}"/>
<Label mi:MauiIcon.Icon="{mi:Cupertino Airplane}"/>
```

**Reply by albyrock87** [reply]
Nope, that doesn't need to be markup extension.

If `BindableObject` doesn't work as base class, we can probably inherit from `Element` like image source does:
https://github.com/dotnet/maui/blob/main/src/Controls/src/Core/ImageSource.cs#L12

The icon value(s) will be set through the attached property `propertyChanged` callback.

**Reply by albyrock87** [reply]
Otherwise if BO doesn't work and we don't like using `Element` we can still use the markup extension and provide `this` as value.

Anyway, I'll look at the PoC once ready.
You can find me on Discord for quick feedbacks.

**Reply by AathifMahir** [reply]
> Otherwise if BO doesn't work and we don't like using `Element` we can still use the markup extension and provide `this` as value.
> 
> Anyway, I'll look at the PoC once ready. You can find me on Discord for quick feedbacks.

Worked on Initial POC, 

What i made sure here is Not Breaking Current Implementation, Therefore I Made the Current Xaml Markup Extension to Complement the New Attached Icon Property, Therefore we are providing the users with both the Classic XamlMarkupExtension and Attached Property

https://github.com/AathifMahir/MauiIcons/tree/Experimental

**Reply by albyrock87** [reply]
Two things:
- `SetBaseIconProperties()` inside the extension shouldn't be needed thanks to `propertyChanged` callback on attached property (as long as you return `this` from the extension)
- the `propertyChanged` callback could propagate the `BindingContext` to the extension so that you can avoid setting it like this `<Setter Property="mi:MauiIcon.Icon" Value="{mi:Cupertino Icon={Binding MyIcon}, BindingContext={x:Reference thisRoot}}" />`

I get that you're trying to not break the API, but that has a cost in terms of code quality, so maybe you can evaluate doing a cleanup in the next major version.

Anyway I think the major problem has been solved ;)

**Reply by AathifMahir** [reply]
> Two things:
> 
> * `SetBaseIconProperties()` inside the extension shouldn't be needed thanks to `propertyChanged` callback on attached property (as long as you return `this` from the extension)
> * the `propertyChanged` callback could propagate the `BindingContext` to the extension so that you can avoid setting it like this `<Setter Property="mi:MauiIcon.Icon" Value="{mi:Cupertino Icon={Binding MyIcon}, BindingContext={x:Reference thisRoot}}" />`
> 
> I get that you're trying to not break the API, but that has a cost in terms of code quality, so maybe you can evaluate doing a cleanup in the next major version.
> 
> Anyway I think the major problem has been solved ;)

1. Would you mind explaining, how we can avoid SetBaseIconProperties Method and InternalBaseSourceProperty, The Problem here is We are returning BindingBase in the MarkupExtension, Therefore we should return new binding instead of BindableObject? in this we can't return this (BaseIconExtension) Since that is a BindableObject not BindingBase. Additionally We can't make the IMarkupExtension Type to an Object Instead of BindingBase, That Would Break Binding or Am i missing something Completely Here?

2. would you mind explaining the How the PropertyChanged Propogate BindingContext to Extension?

Yeah, I'm looking to Gradually Move the Users to New Implementation Since we can use Both the Implementation

Thanks for Reviewing the POC

**Reply by albyrock87** [reply]
I guess you can implement `IMarkupExtension` instead of `IMarkupExtension<T>`.

If that doesn't work (but it should) you can probably create a one time binding to `"."` with `this` as source.

**Reply by AathifMahir** [reply]
> I guess you can implement `IMarkupExtension` instead of `IMarkupExtension<T>`.
> 
> If that doesn't work (but it should) you can probably create a one time binding to `"."` with `this` as source.

I Found that Ideal Solution Would be using SetBaseIconProperties, Since BaseIconExtension is Generic with Enum therefore Converting the Incoming IconExtension into BaseIconExtension is Huge Hassle, We need to have lots of Reflection to Do that.

Additionally, Would you mind explaining how we can propagate BindingContext to the Extension, I'm Pretty Sure that Could be Super Helpful for Many

**Reply by AathifMahir** [reply]
> I guess you can implement `IMarkupExtension` instead of `IMarkupExtension<T>`.
> 
> If that doesn't work (but it should) you can probably create a one time binding to `"."` with `this` as source.

Updated BaseExtension with Improved Implementation, that does get rid of Old SetBaseIconProperties, Instead Passes the Existing Instance to Attached Property,

If you have time, would you be able to review the latest implementation, Thanks

**Reply by albyrock87** [reply]
@AathifMahir would you mind opening a Draft PR so I can comment on specific lines?

**Reply by AathifMahir** [reply]
> @AathifMahir would you mind opening a Draft PR so I can comment on specific lines?

Sure, that works

**Reply by AathifMahir** [reply]
> @AathifMahir would you mind opening a Draft PR so I can comment on specific lines?

Here's the PR
https://github.com/AathifMahir/MauiIcons/pull/108
